lib/OptimizationBBO/src/OptimizationBBO.jl

```diff
@@ -44,16 +46,52 @@ function __map_optimizer_args(prob::SciMLBase.OptimizationProblem, opt::BBO;
         mapped_args = (; mapped_args..., MinDeltaFitnessTolerance=abstol)
     end
 
+    if !isnothing(verbose)
+        mapped_args = (; mapped_args..., TraceMode=verbose)
+    end
+
     return mapped_args
 end
 
+function log_lines(rd, progress, maxiters, id)
+    while isopen(rd)
+        ln = readline(rd)
+        isempty(ln) && continue
+        m = match(r"([0-9]*) steps", ln)
+        if !isnothing(m)
+            s = tryparse(Int, m[1])
+            isnothing(s) && continue
+            isnothing(maxiters) && continue
+            progress && @info ln progress=(s / maxiters) _id=id
+        end
+    end
+end
+
+function redirect_output(f::Function, progress, maxiters, id)
+    old_stdout = stdout
+    rd, = redirect_stdout()
+    task = @async log_lines(rd, progress, maxiters, id)
+    try
+        ret = f()
+        Libc.flush_cstdio()
+        flush(stdout)
+        return ret
+    finally
+        close(rd)
+        redirect_stdout(old_stdout)
+        wait(task)
+    end
+end
+
 function SciMLBase.__solve(prob::SciMLBase.OptimizationProblem, opt::BBO, data=Optimization.DEFAULT_DATA;
```

Comment:
What is `data` and why do we set `maxiters` to `length(data)` if this is provided.

Reply:
It's used for passing a dataset if you want to write a data based loss, for example if you have observations from a system and then write a dynamic model for it and estimate the parameters. Unfortunately it looks like we don't use it much in the docs, we should probably do that more. Minibatching is a good use case for this https://docs.sciml.ai/dev/modules/Optimization/tutorials/minibatch/.  

>why do we set maxiters to length(data) if this is provided.

I think this was a consequence of copy-pasting the optimizationflux implementation, it is done for two reasons, that the callback iterates over the data hence it is required that the number of calls of  it are <= length of data else there will be a index error and second for minibatching where the length is the number of batches.  I agree it should probably not be set to that